File: web_view/src/graphTypes.ts
export interface GraphNode {
  id: string;
  type:
    | 'project'
    | 'directory'
    | 'file'
    | 'namespace'
    | 'struct'
    | 'class'
    | 'union'
    | 'enum'
    | 'function'
    | 'method'
    | 'field'
    | 'constant';
  name: string;
  file: string | null;
  language: string | null;
  line_start: number | null;
  line_end: number | null;
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  type: 'contains' | 'defines' | 'calls' | 'imports' | 'reads' | 'writes';
  source: string;
  target: string;
}

export interface Graph {
  version: string;
  root: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export type ExtensionToWebviewMessage = { type: 'graphData'; graph: Graph };

export type WebviewToExtensionMessage =
  | { type: 'ready' }
  | { type: 'openFile'; relativePath: string };

File: native_analysis_engine/project_scanner/src/graph_types.cpp
#include "codegraph/graph_types.hpp"

namespace codegraph {

// простой генератор
std::string IdGenerator::next() {
    ++counter_;
    return std::to_string(counter_);
}

namespace {

// тут сложно, потому что nlohmann::json не умеет работать с std::optional напрямую
// поэтому мы делаем обертку, которая возвращает либо значение, либо null и тогда nlohmann::json корректно сериализует его в JSON
template <typename T>
nlohmann::json optional_to_json(const std::optional<T>& value) {
    if (value.has_value()) {
        return *value;
    }
    return nullptr;
}

} 
// просто сохраняем в JSON все поля структуры Node, Edge и Graph, используя optional_to_json для полей, которые могут быть пустыми (std::optional)
void to_json(nlohmann::json& j, const Node& n) {
    j = nlohmann::json{
        {"id", n.id},
        {"type", n.type},
        {"name", n.name},
        {"file", optional_to_json(n.file)},
        {"language", optional_to_json(n.language)},
        {"line_start", optional_to_json(n.line_start)},
        {"line_end", optional_to_json(n.line_end)},
        {"metadata", n.metadata},
    };
}

void to_json(nlohmann::json& j, const Edge& e) {
    j = nlohmann::json{
        {"id", e.id},
        {"type", e.type},
        {"source", e.source},
        {"target", e.target},
    };
}

void to_json(nlohmann::json& j, const Graph& g) {
    j = nlohmann::json{
        {"version", g.version},
        {"root", g.root},
        {"nodes", g.nodes},
        {"edges", g.edges},
    };
}

}  
